### PR TITLE
Darlp networks atsc cable virtual channel 1203486985241019

### PR DESCRIPTION
**Pull request by darlp5589**
Hi @triveni-codev , @chriscodev , @pauljustinhernandez 

Kindly check this PR ONLY FOR DIRTY CHANGE and RETRIEVAL/DISPLAY FROM SERVER on both add and update functionalities ONLY FOR ATSC CABLE. 

2 ASANA Tickets associated: 
ASANA: https://app.asana.com/0/1201943683850147/1203486985241019
ASANA: https://app.asana.com/0/1201943683850147/1203486985241021

To test: 
-Virtual Channel should consider the following: 
>1 to 16383 as Major only (no Minor value) is acceptable
>999-999 as biggest Major-Minor value is acceptable
>0-(1 to 999) as Major-Minor values are acceptable
>(1 to 999)-0 as Major Minor values are acceptable
>(1 to 999)-(1 to 999) Major-Minor value combinations are acceptable
>You may now test the BIG COMMIT for both ADDING and UPDATING.

Notes:
0 is not acceptable as Major only value
0-0 is not acceptable as Major-Minor value

Please also test the PATH SELECT and OUT OF BAND checkboxes and make sure they are working.

### CODE REVIEW

**Comment by pauljustinhernandez** [comment]
Hi Darl @darlp5589 , sending here some issues encountered during testing;
1. **0-(1 to 999) as Major-Minor values are acceptable**  ~ As a user I need to press and hold the zero(0) key before I can input dash and minor values.
bug2-need to hold zero

2. **0-(1 to 999) as Major-Minor values are acceptable** ~ greater than 999 (minor value) is accepted when the user press and hold the nine(9) key and move to the other input fields immediately.
bug3-9999 accepted

**Comment by darlp5589** [comment]
Hi @pauljustinhernandez ,

1. 0-(1 to 999) as Major-Minor values are acceptable ~ As a user I need to press and hold the zero(0) key before I can input dash and minor values.
[PLEASE DISREGARD. THIS IS EXACTLY THE SAME BEHAVIOR WITH JAVA APP. IT IS HIGHLY DISCOURAGED TO HAVE 0-(1 to 999) BUT SHOULD BE ACCEPTED IF INTENTIONALLY INPUTTED. YOU CAN ALSO INPUT LIKE 1-999 THEN MANUALLY CHANGE 1 TO 0, THAT IS ACCEPTED IN JAVA APP BUT INTENTIONALLY TYPING 0 FIRST WON'T ALLOW BASED ON THE IMPLEMENTATION]

2. 0-(1 to 999) as Major-Minor values are acceptable ~ greater than 999 (minor value) is accepted when the user press and hold the nine(9) key and move to the other input fields immediately.
[ALREADY FIXED]

File: src/app/_helpers/alphaNumericValidator.ts [hunk 1]
```diff
@@ -15,15 +15,15 @@ export function validateNumberWithDash(event, atscType): boolean {
...
+  if(atscType !== 'ATSC-Terrestrial'){
     majorLimit = 999;
     addByOneIfATSC3 = 1;
   } 
 
   // if atscType is equal to terrestial
   if (atscType === 'ATSC-Terrestrial' || atscType === 'ATSC-3.0') {
     const val = num + event.key;
-    if (!num?.includes('-') && (val < 1 || val > majorLimit)) {
+    if (!num.toString()?.includes('-') && (val < 1 || val > majorLimit)) {
```
Comment (pauljustinhernandez) [inline src/app/_helpers/alphaNumericValidator.ts:26]: ?.toString()

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 2]
```diff
@@ -395,7 +397,7 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
         let major, minor, num; 
         let majorLimit = 99;  
         if(this.modalServiceData.networks.selectedtype === 'ATSC-3.0') majorLimit = 999;   
-        if( this.serviceDataObject.virtual.includes('-')) {
+        if( this.serviceDataObject.virtual.toString().includes('-')) {
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:400]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 3]
```diff
@@ -408,6 +410,25 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
           (error) => console.log('featureService', error)
         );
       }
+      else {
+        if( this.serviceDataObject.virtual.toString().includes('-')) {
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:414]: optional chaining

**Review by pauljustinhernandez — CHANGES_REQUESTED** [review]
Hi Darl @darlp5589 LGTM, I just have comments on your codes before I'll approve on this, mainly~ 
- optional chaining
- commented codes

File: src/app/_helpers/alphaNumericValidator.ts (hunk 1 above)
Comment (darlp5589) [reply src/app/_helpers/alphaNumericValidator.ts:26]: Resolved

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 2 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:400]: Resolved.

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 3 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:414]: Resolved.

File: src/app/pages/main/network-config/networks/networks.component.ts [hunk 4]
```diff
@@ -1010,7 +1010,13 @@ export class NetworksComponent implements OnInit, OnDestroy, OnChanges {
 		// 				if(serviceItem.description === undefined) serviceObject['description'] = "";
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/networks.component.ts:1010]: commented codes; remove if necessary

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 5]
```diff
@@ -452,7 +471,7 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
         if(charCode == 8) { 
           let inputKey = event.target.value;
           let inputKeyLen = inputKey?.length;
-          if(inputKey.includes('-') && inputKey[inputKeyLen-2] != '-') {
+          if(inputKey.toString().includes('-') && inputKey[inputKeyLen-2] != '-') {
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:474]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 6]
```diff
@@ -426,16 +447,14 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
...
             if(this.serviceDataObject.virtual.split('-')[1]?.length > 0) this.isValid = true;
           }
-          else if(inputKey.includes('-') && event.key != '-') this.isValid = true;
+          else if(inputKey.toString().includes('-') && event.key != '-') this.isValid = true;
           else this.isValid = false;
         }
 
         if(this.isValid) {
-          if( inputKey.includes('-')) {
+          if( inputKey.toString().includes('-')) {
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:455]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 7]
```diff
@@ -426,16 +447,14 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
             this.isValid = false;
             if(this.serviceDataObject.virtual.split('-')[1]?.length > 0) this.isValid = true;
           }
-          else if(inputKey.includes('-') && event.key != '-') this.isValid = true;
+          else if(inputKey.toString().includes('-') && event.key != '-') this.isValid = true;
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:450]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 8]
```diff
@@ -426,16 +447,14 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
             this.isValid = false;
             if(this.serviceDataObject.virtual.split('-')[1]?.length > 0) this.isValid = true;
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:448]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 9]
```diff
@@ -948,13 +967,13 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
             return obj;
           });
   
-        if( this.serviceDataObject.virtual.includes('-')) {
-          num = this.serviceDataObject.virtual.split('-');
+        if( this.serviceDataObject.virtual.toString().includes('-')) {
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:970]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 10]
```diff
@@ -426,16 +447,14 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
...
           }
-          else if(inputKey.includes('-') && event.key != '-') this.isValid = true;
+          else if(inputKey.toString().includes('-') && event.key != '-') this.isValid = true;
           else this.isValid = false;
         }
 
         if(this.isValid) {
-          if( inputKey.includes('-')) {
+          if( inputKey.toString().includes('-')) {
             num = inputKey.split('-');
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:456]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 11]
```diff
@@ -408,6 +410,25 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
...
+        if( this.serviceDataObject.virtual?.toString().includes('-')) {
+          let num = this.serviceDataObject.virtual.split('-');
+          let major = num[0];
+          let minor = num[1];
+          if(major === "" || minor === "" || (parseInt(major) === 0 && parseInt(minor) === 0)) this.isValid = false;
+          if(parseInt(major) > 999){
+            major = major.substring(0,3);
+          }  
+          if(parseInt(minor) > 999){
+            minor = minor.substring(0,3);
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:423]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 12]
```diff
@@ -408,6 +410,25 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
...
         );
       }
+      else {
+        if( this.serviceDataObject.virtual?.toString().includes('-')) {
+          let num = this.serviceDataObject.virtual.split('-');
+          let major = num[0];
+          let minor = num[1];
+          if(major === "" || minor === "" || (parseInt(major) === 0 && parseInt(minor) === 0)) this.isValid = false;
+          if(parseInt(major) > 999){
+            major = major.substring(0,3);
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:420]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 13]
```diff
@@ -408,6 +410,25 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
           (error) => console.log('featureService', error)
         );
       }
+      else {
+        if( this.serviceDataObject.virtual?.toString().includes('-')) {
+          let num = this.serviceDataObject.virtual.split('-');
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:415]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 14]
```diff
@@ -408,6 +410,25 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
           (error) => console.log('featureService', error)
         );
       }
+      else {
+        if( this.serviceDataObject.virtual?.toString().includes('-')) {
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:414]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 15]
```diff
@@ -395,7 +397,7 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
         let major, minor, num; 
         let majorLimit = 99;  
         if(this.modalServiceData.networks.selectedtype === 'ATSC-3.0') majorLimit = 999;   
-        if( this.serviceDataObject.virtual.includes('-')) {
+        if( this.serviceDataObject.virtual?.toString().includes('-')) {
           num = this.serviceDataObject.virtual.split('-');
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:401]: optional chaining

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts [hunk 16]
```diff
@@ -395,7 +397,7 @@ export class ModalNetworksAtscServiceComponent implements OnInit, OnChanges, OnD
         let major, minor, num; 
         let majorLimit = 99;  
         if(this.modalServiceData.networks.selectedtype === 'ATSC-3.0') majorLimit = 999;   
-        if( this.serviceDataObject.virtual.includes('-')) {
+        if( this.serviceDataObject.virtual?.toString().includes('-')) {
```
Comment (pauljustinhernandez) [inline src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:400]: optional chaining

**Review by pauljustinhernandez — CHANGES_REQUESTED** [review]
Additonal review Darl, regarding the above latest comment

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 16 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:400]: resolved

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 15 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:401]: resolved

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 13 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:415]: resolved

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 8 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:448]: resolved

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 7 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:450]: resolved

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 6 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:455]: resolved

File: src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts (hunk 5 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/modal-networks/modal-networks-atsc-service/modal-networks-atsc-service.component.ts:474]: resolved

File: src/app/pages/main/network-config/networks/networks.component.ts (hunk 4 above)
Comment (darlp5589) [reply src/app/pages/main/network-config/networks/networks.component.ts:1010]: I am using these entire set of big commit codes time to time on all my PRs. I will remove these in the future.

**Comment by darlp5589** [comment]
Hi @pauljustinhernandez done resolving. You may recheck on them. Thanks!

**Comment by pauljustinhernandez** [comment]
Hi Darl @darlp5589 conducted a final testing for this PR. I encountered an issue ~ see below

1. **1 to 16383 as Major only (no Minor value) is acceptable** ~ greater than 16383 (major only value) is accepted when the user press and hold a desired number key and move to the other input fields immediately. In this example, I input 1638 and press and hold 5.
image

**Comment by darlp5589** [comment]
@pauljustinhernandez , done fixing this concern. You may verify. Thanks!

**Comment by pauljustinhernandez** [comment]
LGTM Darl! @darlp5589

**Pull request merged**